application-admintools-default/src/main/java/com/xwiki/admintools/internal/uploadJob/UploadJobFileProcessor.java

```diff
@@ -0,0 +1,179 @@
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.admintools.internal.uploadJob;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.ZipInputStream;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.AttachmentReference;
+import org.xwiki.model.reference.AttachmentReferenceResolver;
+import org.xwiki.stability.Unstable;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xwiki.admintools.jobs.JobResult;
+import com.xwiki.admintools.jobs.JobResultLevel;
+import com.xwiki.admintools.jobs.PackageUploadJobStatus;
+import com.xwiki.admintools.uploadPackageJob.UploadPackageJobResource;
+
+/**
+ * Processes the {@link XWikiAttachment} and the given archive file content, while also creating a backup if needed.
+ *
+ * @version $Id$
+ * @since 1.1
+ */
+@Component(roles = UploadJobFileProcessor.class)
+@Singleton
+@Unstable
+public class UploadJobFileProcessor
+{
+    /**
+     * Path format for new files from a given parent.
+     */
+    public static final String NEW_FILE_PATH_FORMAT = "%s/%s";
+
+    @Inject
+    private Provider<XWikiContext> xcontextProvider;
+
+    @Inject
+    @Named("current")
+    private AttachmentReferenceResolver<String> attachmentReferenceResolver;
+
+    /**
+     * Checks if a file with the same name or if a JAR file with a similar name exists at the processed target location
+     * and creates a backup if it does.
+     *
+     * @param targetFolderPath the path of the folder for which to check for existing files.
+     * @param filePath the path for the file within the folder.
+     * @param status the job status where to log the backup result.
+     * @return the {@link UploadPackageJobResource} containing the processed target file, file name and backup file.
+     */
+    public UploadPackageJobResource maybeBackupFile(String targetFolderPath, String filePath,
+        PackageUploadJobStatus status)
+    {
+        File targetFile = new File(targetFolderPath + filePath);
+        UploadPackageJobResource uploadPackageJobResource = new UploadPackageJobResource();
+        String fileName = targetFile.getName();
+        uploadPackageJobResource.setNewFileName(fileName);
+        if (targetFile.exists()) {
+            createBackupFile(targetFile.getAbsolutePath() + ".bak", targetFile, uploadPackageJobResource, fileName,
+                status);
+        } else {
+            String extension = filePath.substring(filePath.lastIndexOf("."));
```

Comment:
Could `lastIndexOf` return -1? (if there is a file without extension) In this case, the `substring` method will throw an IndexOutOfBounds exception.